aoetools-36/aoecfg.8
.TH aoecfg 8
.SH NAME
aoecfg \- manipulate AoE configuration strings
.SH SYNOPSIS
.B aoecfg
[-c \fIcmd\fR] [-s \fIcfgstr\fR] [-t \fItimeout\fR] [\fIshelf slot\fR] [\fInetif\fR]
.fi
.SH DESCRIPTION
.IR Aoecfg (8)
sends AoE configuration commands that control the retrieval, conditional
or unconditional setting of AoE configuration strings.  Since configuration
happens before the MAC address of the target is known, the packet is
broadcast.  AoE targets with a matching shelf and slot respond.  Since
the default shelf and slot are the wildcard values 0xffff and 0xff,
with no arguments
.IR aoecfg (8)
will return configuration strings from all targets visible
on the default interface,
.IR eth0 .
.SH OPTIONS
.TP 8
.BI \-c " cmd"
specify the AoE configuration command.  The default is
.IR read .
The available commands are
.HP 8
.B read
Read the server config string without performing any test and
respond.
.HP 8
.B test
Respond only if the specified string exactly matches the server
configuration string.
.HP 8
.B prefix
Respond only if the specified string is a prefix of the server
configuration string.
.HP 8
.B set
If the current server config string is empty, set the server config
string to the argument string and respond.  If the current server
config string is not empty, return a response with Flags bit E set
and Error set to 4.
.HP 8
.B fset
Force set the server config string to the argument string and respond.
.TP
.BI \-s " cfgstr"
specify the config string.
.TP
.BI \-t " timeout"
specify the timeout in seconds.  The default is no timeout.  If neither the shelf
nor the slot are specified,
.IR aoecfg (8)
will exit after the first result.  Otherwise,
.IR aoecfg (8)
will exit only after the timeout has expired since it does not know
how many responses to expect.
.TP
.B shelf slot
specify the shelf and slot used in the query.  If unspecified, they
default to broadcast.
.TP
.B netif
specify the network interface.  The default is
.IR eth0 .
.SH "SEE ALSO"
.IR aoe-discover (8),
.IR aoe-interfaces (8),
.IR aoe-mkdevs (8),
.IR aoe-mkshelf (8),
.IR aoe-stat (8),
.IR aoeping (8),
\fIAoE (ATA over Ethernet)\fP: http://support.coraid.com/documents/AoEr10.txt,
\fIATA specification\fP
.SH AUTHOR
Erik Quanstrom (quanstro@coraid.com)


NEWS
-*- change-log -*-
2017-03-15 Ed Cashin <ed.cashin@acm.org>
	mothballing standalone driver
	version-v87moth

2017-02-22 Ed Cashin <ed.cashin@acm.org>
	merge v87pre3-bi_error for 4.4.x support
	version-87

2016-03-16 Ed Cashin <ed.cashin@acm.org>
	include "etherd/" in character device names for udev-systemd
	version-87pre3

2016-01-07 Ed Cashin <ed.cashin@acm.org>
	update driver version string in aoe.h
	version-87pre2

2015-12-13 Ed Cashin <ed.cashin@acm.org>
	correct conf/41 kernel version test (immutable biovec support)
	version-87pre1

2015-07-04 Ed Cashin <ed.cashin@acm.org>
	version-86

2015-04-13 Ed Cashin <ed.cashin@acm.org>
	add support for immutable biovecs with backward compatibility
	version-86pre1

2015-04-01 Ed Cashin <ed.cashin@acm.org>
	update documentation with support.coraid.com offline

2013-10-14 Sam Hopkins <sah@coraid.com>
	version-85

2013-09-17 Ed Cashin <ecashin@coraid.com>
	fix compat problem with building on Fedora 19

2013-09-17 Ed Cashin <ecashin@coraid.com>
	version-85pre1

2013-09-10 Ed Cashin <ecashin@coraid.com>
	use aoetools-36

2013-08-26 Ed Cashin <ecashin@coraid.com>
	update compatibility system
	drop support for kernels older than 2.6.18

2013-08-22 Ed Cashin <ecashin@coraid.com>
	get page head reference for compound page tails in biovecs
	remove warning-generating do-nothing term from example udev rules

2013-07-10 Ed Cashin <ecashin@coraid.com>
	add pkg subdirectory for distro-specific packaging files
	version-84

2013-07-02 Ed Cashin <ecashin@coraid.com>
	merge upstream patch: reservation of MAX_HEADER bytes in skbs
	version-84pre3

2013-06-19 Ed Cashin <ecashin@coraid.com>
	converge style (whitespace, etc.) towards kernel.org-distributed aoe
	version-84pre2

2013-06-04 Ed Cashin <ecashin@coraid.com>
	note in README: 2.6.16 will be oldest supported kernel
	version-84pre1

2013-05-26 Ed Cashin <ecashin@coraid.com>
	upstream change: use kmemdup when appropriate
	move code snippets to match order in kernel.org

2013-05-21 Ed Cashin <ecashin@coraid.com>
	do not use skb members after calling dev_queue_xmit
	update URLs in README

2013-02-11 Swanand Rao <swrao@coraid.com>
	fix bug where rmmod causes driver to panic
	version-83pre1
	version-83

2012-12-13 Ed Cashin <ecashin@coraid.com>
	fix hard-to-hit use-after-free bug found by Dan Carpenter

2012-10-30 Swanand Rao <swrao@coraid.com>
	add support for parallelization of I/O completion
	version-82pre2
	version-82

2012-09-20 Ed Cashin <ecashin@coraid.com>
	clean up kernel thread management routine
	use aoetools-35

2012-08-15 Ed Cashin <ecashin@coraid.com>
	revert get_page use; avoid using compound page tail for zero page
	fix aoe_ktio pausing of version 80
	version-81pre2
	version-81

2012-08-03 Ed Cashin <ecashin@coraid.com>
	use {get,put}_page for page _count, avoiding panic for compound pages
	version-81pre1

2012-07-17 Ed Cashin <ecashin@coraid.com>
	version-80

2012-06-29 Ed Cashin <ecashin@coraid.com>
	use aoetools-34
	cap I/O completion iterations before yeilding CPU in aoe_ktio
	support 3.2.y kernel
	version-80pre1

2012-02-09 Ed Cashin <ecashin@coraid.com>
	version-79

2012-01-27 Ed Cashin <ecashin@coraid.com>
	tell Linux virtual memory subsystem when aoe_tx writes
	version-79pre4

2011-12-30 Ed Cashin <ecashin@coraid.com>
	use aoetools-33
	update virtual memory settings information in HOWTO
	version-79pre3

2011-12-22 Ed Cashin <ecashin@coraid.com>
	avoid compiler warning in openSUSE 11.4
	update compat system: 1-param __blk_run_queue is latest
	version-79pre2

2011-12-14 Ed Cashin <ecashin@coraid.com>
	support openSUSE 11.4 in compatibility system
	version-79pre1

2011-12-08 Ed Cashin <ecashin@coraid.com>
	support 2.6.39 and newer kernels without big kernel lock
	version-78pre2
	version-78

2011-07-19 Ed Cashin <ecashin@coraid.com>
	support Scientific Linux 6.0 with check for build dependency: perl
	version-78pre1

2011-05-25 Ed Cashin <ecashin@coraid.com>
	support 2.6.38
	version-77

2010-11-19 Ed Cashin <ecashin@coraid.com>
	use aoetools-32
	avoid sending more than one probe at a time on tainted path
	get more I/O to do after stopping probes
	do not probe unless untainted aoetgts exist
	support 2.6.36
	version-76

2010-06-21 Ed Cashin <ecashin@coraid.com>
	support new name for blk_queue_max_sectors in 2.6.34
	get kmalloc and kfree declarations from slab.h to support 2.6.34
	version-75

2010-06-09 Ed Cashin <ecashin@coraid.com>
	avoid using system time in coarse-grained calculations
	support include/asm symlink removal in 2.6.33
	support SLES 10's unusual placement of skb_reset_mac_header

2010-01-08 Ed Cashin <ecashin@coraid.com>
	update copyright notices
	version-74

2009-11-23 Ed Cashin <ecashin@coraid.com>
	impose cap on retransmit time to single remote MAC address
	remove memory leak in grow_targets for more than 4 remote MACs
	ignore outstanding probes in decision to call I/O request handler
	only use one probe per tainted remote MAC address

2009-09-21 Ed Cashin <ecashin@coraid.com>
	do not run autoload.sh to configure loading of aoe on dev node access
	use aoetools-30
	support changes in 2.6.31
	version-73

2009-07-20 Ed Cashin <ecashin@coraid.com>
	use ATA_CMD_ values from ata.h instead of WIN_ macros in hdreg.h
	support older kernels where ata.h is incompatible with hdreg.h
	allow aoedbg.c to see DEFINE_SPINLOCK in pre 2.6.11 kernels
	version-72

2009-06-29 Ed Cashin <ecashin@coraid.com>
	do not use relay when Module.symvers is not present

2009-04-09 Ed Cashin <ecashin@coraid.com>
	avoid premature I/O failure: maintain timings on frame reassignment
	version-71

2009-02-25 Ed Cashin <ecashin@coraid.com>
	support commas instead of spaces in aoe_iflist parameter
	Roel Kluin: in skbfree, use skb if it's available on last iteration

2009-02-06 Ed Cashin <ecashin@coraid.com>
	fix the "kerninst" target for make, which had atrophied

2009-02-03 Ed Cashin <ecashin@coraid.com>
	fix and refine compatibility tests
	version-70

2009-01-28 Ed Cashin <ecashin@coraid.com>

aoetools-36/aoecfg.c
/*
 * aoecfgstr.c - fiddle aoe configuration strings.
 * Copyright 2009, Erik Quanstrom, CORAID, Inc., Licenced under GPL v2
 */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdint.h>
#include <ctype.h>
#include <netinet/in.h>
#include <errno.h>
#include <sys/time.h>
#include <time.h>
#include <arpa/inet.h>
#include "dat.h"
#include "fns.h"

int	sfd;
uchar	mac[6];
int	timeout;
u16	shelf	= 0xffff;
uchar	slot	= 0xff;
char	*net 	= "eth0";
char	*cfgstr;
int	cfgstrlen;

char *errtab[7] = {
	"*GOK*",
	"*badcmd*",
	"*badarg*",
	"*baddev*",
	"*badcfg*",
	"*badvers*",
	"*GOK*",
};

void
resp(Conf *c)
{
	Aoehdr *h;
	char *s;
	int l;

	h = (Aoehdr*) c;
	if (h->flags & Error) {
		s = errtab[h->error & 7];
		l = strlen(s);
	} else {
		s = (char*) c->data;
		l = ntohs(c->len);
	}
	if (shelf != 0xffff && slot != 0xff)
		printf("%.*s\n", l, s);
	else
		printf("%d.%d\t%.*s\n", ntohs(h->maj), h->min, l, s);
}

int
readto(int fd, void *buf, int size)
{
	fd_set rfd;
	struct timeval tv;
	static int to;

	if (timeout == 0)
		goto f1;
	if (to == 0)
		to = time(0) + timeout;
	FD_ZERO(&rfd);
	FD_SET(fd, &rfd);

	tv.tv_sec = to - time(0);
	tv.tv_usec = 0;

	switch (select(fd+1, &rfd, 0, 0, &tv)) {
	case -1:
		perror("select");
		exit(1);
	case 0:
		exit(0);
	}
f1:
	return read(fd, buf, size);
}

u32
aoe_tag(void)
{
	u32 n;
	struct timeval t = { 0, 0 };

	/* Tag should be just enough to avoid conflicts with other
	 * aoeping and aoecfg processes, and should set high bit
	 * to avoid conflicting with in-kernel AoE.
	 */
	if (gettimeofday(&t, NULL)) {
		perror("gettimeofday");
		exit(1);
	}
	n = t.tv_usec | 1UL << 31;
	return htonl(n);
}

void
cfgquery(int cmd, int shelf, int slot)
{
	int n;
	Aoehdr *h;
	Conf *c;
	uchar buf[1024 + sizeof *h];
	u32 tag;

	c = (Conf*) buf;
	h = (Aoehdr*) c;
	memset(h, 0, sizeof *h);
	memset(h->dst, 0xff, sizeof h->dst);
	memmove(h->src, mac, sizeof h->src);
	
	h->type = htons(AOE_ETH_PROTO);
	h->flags = AoEver << 4;
	h->maj = htons(shelf);
	h->min = slot;
	h->cmd = Config;
	tag = aoe_tag();
	memmove(h->tag, &tag, sizeof h->tag);
	c->bufcnt = 0;
	c->firmware = 0;
	c->scnt = 0;
	c->vercmd = cmd;
	memmove(c->data, cfgstr, cfgstrlen);
	c->len = htons(cfgstrlen);
	if (write(sfd, c, sizeof *c) == -1) {
		perror("send config query");
		exit(1);
	}
	for (;;) {
		n = readto(sfd, buf, sizeof buf);
		if (n < 0) {
			perror("read network");
			exit(1);
		}
		if (n < 60)
			continue;
		h = (Aoehdr *) buf;
		if (ntohs(h->type) != AOE_ETH_PROTO)
			continue;
		if (ntohs(h->maj) == 0xffff || h->min == 0xff)
			continue;
		if (shelf != 0xffff && ntohs(h->maj) != shelf)
			continue;
		if (slot != 0xff && h->min != slot)
			continue;
		if (memcmp(h->tag, &tag, sizeof h->tag))
			continue;
		resp((Conf*) buf);
		if (shelf != 0xffff && slot != 0xff)
			break;
	}
}

void
usage(void)
{
	fputs("usage: aoecfg "
		"[-c cmd] [-s cfgstr] [-t timeout] "
		"[shelf slot] [net]\n", stderr);
	exit(1);
}

typedef struct{
	char	*s;
	int	cmd;
} Tab;

Tab tab[] = {
	{ "read",	Qread, },
	{ "test",	Qtest, },
	{ "prefix",	Qprefix, },
	{ "set",	Qset, },
	{ "fset",	Qfset, },
};

int
xlatecmd(char *s)
{
	int i;

	for (i = 0; i < nelem(tab); i++)
		if (strcmp(tab[i].s, s) == 0)
			return tab[i].cmd;
	usage();
	return -1;	// whine whine whine
}

int
main(int argc, char *argv[])
{
	int c, cmd;

	cmd = Qread;
	while ((c = getopt(argc, argv, "s:c:t:v")) != -1) {
		switch (c) {
		case 'c':
			cmd = xlatecmd(optarg);
			break;
		case 's':
			cfgstr = optarg;
			cfgstrlen = strlen(cfgstr);
			break;
		case 't':
			timeout = atoi(optarg);
			break;
		default:
			usage();
		}
	}

	c = argc - optind;
	argv += optind;
	if (c < 2)
		goto f1;
	if (c != 3 && c != 2)
		usage();
	shelf = strtoul(*argv++, 0, 0);
	slot = strtoul(*argv++, 0, 0);
f1:
	if (*argv) {
		net = *argv;
		if (isdigit((int) *net)) {
			fprintf(stderr,
				"aoecfg Error: \"%s\" %s\n", net,
				"is not a valid network interface");
			usage();
		}
	}
	sfd = dial(net);
	if (getea(sfd, net, mac) == 0)
		exit(1);
	cfgquery(cmd, shelf, slot);
	return 0;
}
